gradle/docker.gradle.kts

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2019 Knot.x Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import com.bmuschko.gradle.docker.tasks.image.DockerBuildImage
+import com.bmuschko.gradle.docker.tasks.image.DockerRemoveImage
+import com.bmuschko.gradle.docker.tasks.image.DockerSaveImage
+
+val dockerBaseImageRef = "$buildDir/.docker/buildBaseImage-imageId.txt"
+val dockerBaseAlpineImageRef = "$buildDir/.docker/buildBaseAlpineImage-imageId.txt"
+
+val dockerBaseImageId = "${project.property("docker.domain")}/knotx:${project.version}"
+val dockerBaseAlpineImageId = "${project.property("docker.domain")}/knotx-alpine:${project.version}"
+
+val dockerfileBaseImagePath = "$projectDir/src/main/docker/base/Dockerfile"
+val dockerfileBaseAlpineImagePath = "$projectDir/src/main/docker/base-alpine/Dockerfile"
+
+fun DockerRemoveImage.remove(name: String): Unit {
+    val spec = object : Spec<Task> {
+        override fun isSatisfiedBy(task: Task): Boolean {
+            return File(name).exists()
+        }
+    }
+    onlyIf(spec)
+
+    targetImageId(if (File(name).exists()) File(name).readText() else "")
+    onError {
+        if (!this.message!!.contains("No such image"))
+            throw this
+    }
+}
+
+tasks.register<Copy>("copyDockerfileBase") {
+    group = "docker"
+    from(dockerfileBaseImagePath)
+    into("$buildDir/out/base")
+}
+
+tasks.register<Copy>("copyDockerfileBaseAlpine") {
+    group = "docker"
+    from(dockerfileBaseAlpineImagePath)
+    into("$buildDir/out/baseAlpine")
+}
+
+tasks.register<DockerRemoveImage>("removeBaseImage") {
+    group = "docker"
+    remove(dockerBaseImageRef)
+}
+
+tasks.register<DockerRemoveImage>("removeBaseAlpineImage") {
+    group = "docker"
+    remove(dockerBaseAlpineImageRef)
+}
+
+tasks.register<DockerBuildImage> ("buildBaseImage") {
+    group = "docker"
+    inputDir.set(file("$buildDir/out"))
+    dockerFile.set(file("$buildDir/out/base/Dockerfile"))
+    images.add(dockerBaseImageId)
+    dependsOn("removeBaseImage", "copyDockerfileBase")
+}
+
+tasks.register<DockerBuildImage> ("buildBaseAlpineImage") {
+    group = "docker"
+    inputDir.set(file("$buildDir/out"))
+    dockerFile.set(file("$buildDir/out/baseAlpine/Dockerfile"))
+    images.add(dockerBaseAlpineImageId)
+    dependsOn("removeBaseAlpineImage", "copyDockerfileBaseAlpine")
+}
+
+tasks.register<DockerSaveImage>("saveBaseImage") {
```

Comment:
I am wondering if we can:
- define `base` and `alpine` as variations of tasks, what if we add some `example` in the future
- then we can define variations' list and register tasks groups per variation

Reply:
I was thinking the same, `remove` task has logic unified into the Kotlin method.
However, tasks like `saveImage` or `buildImage` have every configuration value different, so even when we abstract their logic, still there will be the same numbers of params to configure.
I came to the conclusion that this form is easier to understand.
Do you see any particular example to extract logic to the more general task/method?